### ISSUE TITLE
Question about Bounding Box

### ISSUE THREAD

**Issue by SwcK423** [nerficg-project/SPaGS#5]
I try to visualize the 3d Gaussian, but when the GS is closed to camera or has a large scale, the calculation of bbox may be failed. It manifests as missing certain areas or large areas, or even no coverage at all. Even though I don't think there are any problems in your mathematics, could there be certain deficiencies in the code?

**Comment by JunboLi-CN** [comment]
Hello,

thank you for your interest in our work. I hadn't noticed the problem before you mentioned it. Could you please check whether it is due to your near-far plane parameter settings or whether other culling conditions have been reached? Otherwise, there is one more thing that may be relevant: when the camera center is inside the 3D Gaussian, it will also be culled. You may have missed this if the Gaussian is too big. If none of the above applies, could you please provide a short video showing the failure case? It would be much appreciated!

Best Regards

**Comment by JunboLi-CN** [comment]
By the way, which version of the Nerficg framework do you use? There have been a few updates to Nerficg and other dependencies since this code was released, which may cause some compatibility issues. The current recommended version of Nerficg as the dependency for this code is the commit: fda9938c7e64b6605562f79d696c11baff52186e, and all other dependencies should be set to the nearest older commit time as this code. To resolve the dependencies issue I will update the code once I have more time. Sorry for the inconvenience. Please try the above suggestions first, and then let me know if they solve your problem.

**Comment by SwcK423** [comment]
<img width="1182" height="782" alt="Image" src="https://github.com/user-attachments/assets/9580fc1d-83d2-4f3c-b721-77bea6e20b4d" />

near_plane=0.001, I tried to mask each if-return term in transform_and_cull, it still failed as illustration.

**Comment by JunboLi-CN** [comment]
Hello, 

thank you for sharing the test video. It seems like the code may really have an issue with BBX on the Y-axis. Would you mind providing your test code and test parameter settings so that I can quickly reproduce the issue? It would be much appreciated!

**Comment by SwcK423** [comment]
import torch
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider, CheckButtons
from scipy.spatial.transform import Rotation as R

from submodules.SPaGSCudaBackend.SPaGSCudaBackend.torch_bindings.rasterization import SPaGSRasterizer


def start_rasterizer_ui():
    W, H = 1024, 512
    device = torch.device("cuda")

    viewmatrix = torch.eye(4, device=device)
    campos = torch.zeros(3, device=device)

    # UI 布局
    fig, ax = plt.subplots(figsize=(14, 8))
    plt.subplots_adjust(bottom=0.35, right=0.85)
    img_display = ax.imshow(np.zeros((H, W, 3)), origin='upper', extent=[0, W, H, 0], aspect='auto')

    # 滑块控制 (控制单个高斯点作为测试)
    s_cfg = [
        ('X', -10, 10, 0), ('Y', -10, 10, 2), ('Z', -10, 10, 5),
        ('Scale_X', 0.1, 10, 1.5), ('Scale_Y', 0.1, 10, 0.5), ('Scale_Z', 0.0, 10, 0.5),
        ('Pitch', -3.14, 3.14, 0.0), ('Yaw', -3.14, 3.14, 0.0)
    ]
    sliders = {c[0]: Slider(plt.axes([0.15, 0.26 - i * 0.035, 0.6, 0.02]), c[0], c[1], c[2], valinit=c[3])
               for i, c in enumerate(s_cfg)}
    patches = []

    def update(val):
        nonlocal patches
        for p in patches: p.remove()
        patches.clear()

        mu = np.array([sliders['X'].val, sliders['Y'].val, sliders['Z'].val])
        sc = np.array([sliders['Scale_X'].val, sliders['Scale_Y'].val, sliders['Scale_Z'].val])
        rot_obj = R.from_euler('xy', [sliders['Pitch'].val, sliders['Yaw'].val])

        means3D = torch.from_numpy(mu).float().cuda().unsqueeze(0)
        scales = torch.from_numpy(sc).float().cuda().unsqueeze(0)
        rotations = torch.from_numpy(rot_obj.as_quat()).float().cuda().unsqueeze(0)
        opacities = torch.tensor([[0.7]], device=device)

        shs = torch.tensor([[0.2, 0.8, 0.5]], device=device)
        shs_rest = torch.zeros((1, 3, 15), device=device)

        rasterizer = SPaGSRasterizer()

        with torch.no_grad():
            image, depth = rasterizer(
                positions=means3D,
                scales=scales,
                rotations=rotations,
                opacities=opacities,
                sh_0=shs,
                sh_rest=shs_rest,
                densification_info=torch.empty(0),
                M=viewmatrix.transpose(0, 1),
                cam_position=campos,
                width=int(W),
                height=int(H),
                near=0.001,
                mode=0,
                K=16,
                active_sh_bases=0,
                scale_modifier=1.0,
                use_distance_scaling=False,
            )

            img_np = image.detach().cpu().permute(1, 2, 0).numpy()

        img_display.set_data(np.clip(img_np, 0, 1))

        fig.canvas.draw_idle()

    for s in sliders.values(): s.on_changed(update)
    update(None)
    plt.show()

start_rasterizer_ui()

**Comment by JunboLi-CN** [comment]
Hello,

sorry for the delayed reply — I've been rather busy today. I have just tried your test code and, strangely, I didn't encounter the same issue as you. I tried with the same parameter settings as yours, as you can see in the accompanying video screenshots. Did you change our source code at any point when you tried it? You could try cloning the code and running the test again to see if the issue is due to any changes you have made.

https://github.com/user-attachments/assets/a83e1ecf-d3a0-4f29-96cf-c2337e46b933

https://github.com/user-attachments/assets/015f7933-7904-489b-84af-bbe217f2b4d3

https://github.com/user-attachments/assets/abeccfb3-e2e4-4f14-9130-4b22078fdb7d

The Gaussian may disappear in between. That's normal due to the culling I mentioned earlier. Just try moving it away from the camera slightly and it should reappear (it also works if you want to make the scale larger).

**Comment by JunboLi-CN** [comment]
By the way, I have adapted your test code slightly to avoid changing the interface of our rasterizer.

```python
from Cameras.Perspective import PerspectiveCamera
from Cameras.utils import CameraProperties

def start_rasterizer_ui():
    W, H = 1024, 512
    device = torch.device("cuda")

    viewmatrix = torch.eye(4, device=device)
    campos = torch.zeros(3, device=device)
    cam_near = 0.01
    cam_far = 1000.0
    cam_properties = CameraProperties()
    cam_properties.width = int(W)
    cam_properties.height = int(H)
    cam_properties.c2w = viewmatrix # camera in (0, 0, 0) looking towards -z
    #cam_properties.c2w[:3, 3] = campos # is the same as above since campos is (0,0,0)
    cam = PerspectiveCamera(cam_near, cam_far)
    cam.properties = cam_properties

    # UI 布局
    fig, ax = plt.subplots(figsize=(14, 8))
    plt.subplots_adjust(bottom=0.35, right=0.85)
    img_display = ax.imshow(np.zeros((H, W, 3)), origin='upper', extent=[0, W, H, 0], aspect='auto')

    # 滑块控制 (控制单个高斯点作为测试)
    s_cfg = [
        ('X', -10, 10, 0), ('Y', -10, 10, 1.29), ('Z', -10, 10, 1.74),
        ('Scale_X', 0.1, 10, 3.235), ('Scale_Y', 0.1, 10, 0.748), ('Scale_Z', 0.0, 10, 0),
        ('Pitch', -3.14, 3.14, 0.015), ('Yaw', -3.14, 3.14, 0.022)
    ]
    sliders = {c[0]: Slider(plt.axes([0.15, 0.26 - i * 0.035, 0.6, 0.02]), c[0], c[1], c[2], valinit=c[3])
            for i, c in enumerate(s_cfg)}
    #patches = []

    def update(val):
        # nonlocal patches
        # for p in patches: 
        #     p.remove()
        #     patches.clear()

        mu = np.array([sliders['X'].val, sliders['Y'].val, sliders['Z'].val])
        sc = np.array([sliders['Scale_X'].val, sliders['Scale_Y'].val, sliders['Scale_Z'].val])
        rot_obj = R.from_euler('xy', [sliders['Pitch'].val, sliders['Yaw'].val])

        means3D = torch.from_numpy(mu).float().cuda().unsqueeze(0)
        scales = torch.from_numpy(sc).float().cuda().unsqueeze(0)
        rotations = torch.from_numpy(rot_obj.as_quat()).float().cuda().unsqueeze(0)
        opacities = torch.tensor([[0.7]], device=device)

        shs = torch.tensor([[0.2, 0.8, 0.5]], device=device)
        shs_rest = torch.zeros((1, 3, 15), device=device)

        rasterizer = SPaGSRasterizer()

        with torch.no_grad():
            image = rasterizer(
                positions=means3D,
                scales=scales,
                rotations=rotations,
                opacities=opacities,
                sh_0=shs,
                sh_rest=shs_rest,
                densification_info=torch.empty(0),
                #M=viewmatrix.transpose(0, 1),
                #cam_position=campos,
                #width=int(W),
                #height=int(H),
                #near=0.001,
                camera=cam,
                mode=0,
                K=16,
                active_sh_bases=0,
                scale_modifier=1.0,
                use_distance_scaling=False,
            )

            img_np = image.detach().cpu().permute(1, 2, 0).numpy()

        img_display.set_data(np.clip(img_np, 0, 1))

        fig.canvas.draw_idle()

    for s in sliders.values(): 
        s.on_changed(update)
        update(None)
    plt.show()
```

**Comment by SwcK423** [comment]
Hello,

Thanks for reply, 
I tried to mask some codes in previous visualization:
if ((center_tangent.z + extent_tangent.z) >= 1.0f || (center_tangent.z - extent_tangent.z) <= -1.0f) return true;
if ((extent_tangent.x < BB_EPS) || (extent_tangent.y < BB_EPS) || (extent_tangent.z < BB_EPS)) return true;
if ((min_bounds.x >= 0.0f) || (min_bounds.y >= 0.0f) || (max_bounds.x <= 0.0f) || (max_bounds.y <= 0.0f)) return true;
if (px_max < px_min) return true;
if ((px_max - px_min < 1.0f) || (py_max - py_min < 1.0f)) return true;

Although I use your original codes, it still remains some problems like:

<img width="1400" height="800" alt="Image" src="https://github.com/user-attachments/assets/e09b4eb7-1a5b-4c10-8d5f-cb5b4655ad63" />

Is it correct that gaussian disappears each time in the video? In many cases, the sampling method still works properly.

**Comment by JunboLi-CN** [comment]
Hello,

yes, what is shown in your videos is normal. According to the parameter settings of your test code, the camera model in the world coordinate system should look like this (the red line is the edge of the equirectangular image):

<img width="1222" height="1203" alt="Image" src="https://github.com/user-attachments/assets/6e75c0c1-d938-4dfb-8757-c886430cfaf0" />

RGB -> XYZ. With this, you can easily determine the relative pose relationship between the camera and the Gaussian, and identify when it will overlap with the near plane or the camera centre, which triggers the culling. I hope this helps. Have a great day!

Best Regards

**Comment by SwcK423** [comment]
<img width="1400" height="800" alt="Image" src="https://github.com/user-attachments/assets/5f414a21-59ac-4c18-a087-119f2a4eaf5a" />

The former employ BBX, the latter remove it. Is there any situation of unstable numerical precision in the calculation of the if statement?

**Comment by JunboLi-CN** [comment]
Which situation do you mean? The BBX is definitely accurate, I suppose your question is just when it should be culled and when not, right?

**Comment by SwcK423** [comment]
Including the situations previously mentioned where bbox disappeared or was not fully covered. I think these situations should actually be included in the subsequent calculations rather than being culled. Of course, the most obvious case is that when I mask all the if-cull judgment, the bounding box coverage is still incomplete.

**Comment by JunboLi-CN** [comment]
I think you have misunderstood the purpose of the culling. In this field, culling is quite commonly used to handle situations that are physically and/or mathematically impossible, e.g. the camera in a Gaussian or a Gaussian occurring behind the image plane. Such situations directly contradict the theoretical basis of this work, such as theoretical camera models and intersection test. Therefore, you cannot simply disable all culling mechanisms and still expect everything to work perfectly. If you are looking for a solution to these situations, it is beyond the scope of this work and you will probably need to use different theoretical model.

**Comment by SwcK423** [comment]
OK, thanks for your reply!
